### PR TITLE
Restyling - new login screen

### PR DESCRIPTION
**Pull request by dcluna**
Redesigned login screen

### CODE REVIEW

File: app/res/drawable/login_edit_text.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <padding>8dp</padding>
+    <size android:height="48dp"></size>
```
Comment (ctsims) [inline app/res/drawable/login_edit_text.xml]: These are standard constants for us now, right? They should be either in a style or dimensions file, shoudln't they?

File: app/res/layout/screen_login.xml [hunk 2]
```diff
@@ -1,48 +1,56 @@
...
         android:layout_alignParentTop="true"
         android:orientation="vertical" >
 
         <FrameLayout
             android:layout_width="fill_parent"
-            android:layout_height="0px" android:layout_weight=".2" android:id="@+id/screen_login_banner_pane">
+            android:layout_height="0px"
+            android:layout_weight=".2"
+            android:paddingTop="8dp"
+            android:paddingLeft="8dp"
```
Comment (ctsims) [inline app/res/layout/screen_login.xml]: I think these 8dp's should also be in either a standard style or in dimens.xml

File: app/res/layout/screen_login.xml [hunk 3]
```diff
@@ -1,48 +1,56 @@
...
         android:layout_height="wrap_content"
-        android:layout_above="@id/screen_login_bottom_pane"
         android:layout_alignParentTop="true"
         android:orientation="vertical" >
 
         <FrameLayout
             android:layout_width="fill_parent"
-            android:layout_height="0px" android:layout_weight=".2" android:id="@+id/screen_login_banner_pane">
+            android:layout_height="0px"
+            android:layout_weight=".2"
```
Comment (ctsims) [inline app/res/layout/screen_login.xml]: do we think this is always the right weight? might be useful to think about a max_height or min_height depending on the screen size.

File: app/res/values/styles.xml [hunk 4]
```diff
@@ -137,4 +137,33 @@
         <item name="android:textColor">@color/white</item>
     </style>
 
-</resources>
\ No newline at end of file
+    <style name="ListViewV2">
```
Comment (ctsims) [inline app/res/values/styles.xml]: Is there a better descriptor than V2 for these? Most of these before didn't actually have V1's, did they? I'm fine with just naming them, although context would also be helpful if they're only used in one specific place currently

File: app/src/org/commcare/dalvik/activities/LoginActivity.java [hunk 5]
```diff
@@ -52,14 +55,14 @@
...
-    
+
     @UiElement(value=R.id.login_button, locale="login.button")
     Button login;
     
-    @UiElement(value=R.id.text_username, locale="login.username")
-    TextView userLabel;
-    @UiElement(value=R.id.text_password, locale="login.password")
-    TextView passLabel;
+//    @UiElement(value=R.id.text_username, locale="login.username")
```
Comment (ctsims) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: Can just delete these entirely.

File: app/res/values/styles.xml (hunk 4 above)
Comment (dcluna) [inline app/res/values/styles.xml]: Maybe we should hold renaming (and doing other small code fixes) until we merge everything else? This style is used in a lot of places in the redesigned code, and I'd prefer if we avoid changing it for now.

File: app/src/org/commcare/dalvik/activities/LoginActivity.java [hunk 6]
```diff
@@ -87,6 +110,7 @@
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         username.setInputType(InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS | InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD);
+        editTextColor = username.getCurrentTextColor();
```
Comment (wpride) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: recovering a color from a UI element seems like an odd pattern - are we basically using this to keep track of normal vs. red after bad login?

File: app/res/layout/screen_login.xml [hunk 7]
```diff
@@ -51,60 +59,91 @@
...
+                    style="@style/LoginEditTextV2"
+                    android:hint="Password"
+                    android:textSize="@dimen/text_medium"
+                    />
+
+                <TextView
+                    android:id="@+id/screen_login_bad_password"
+                    android:layout_width="fill_parent"
+                    android:layout_height="wrap_content"
+                    android:text="We couldn't find a user with this password. Please try another!"
```
Comment (wpride) [inline app/res/layout/screen_login.xml]: Is this localizable somewhere?

File: app/res/layout/screen_login.xml [hunk 8]
```diff
@@ -51,60 +59,91 @@
...
+                <EditText
+                    android:id="@+id/edit_username"
+                    android:layout_width="fill_parent"
+                    android:layout_height="wrap_content"
+                    android:inputType="text"
+                    android:nextFocusDown="@+id/edit_password" android:contentDescription="Username"
+                    android:drawableLeft="@drawable/icon_user_neutral50"
+                    android:background="@drawable/login_edit_text"
+                    style="@style/LoginEditTextV2"
+                    android:hint="Username"
```
Comment (wpride) [inline app/res/layout/screen_login.xml]: What's the method for localizing hints?

File: app/res/layout/screen_login.xml [hunk 9]
```diff
@@ -51,60 +59,91 @@
...
 
-            <TextView
-                android:id="@+id/screen_login_bad_password"
-                android:layout_width="fill_parent"
-                android:layout_height="wrap_content"
-                android:text="Incorrect Username or Password" android:background="@drawable/bubble_danger" android:paddingTop="2dp" android:paddingBottom="2dp" android:textColor="@color/red" android:textStyle="bold" android:visibility="gone" android:layout_marginLeft="2dp" android:paddingLeft="5dp" android:paddingRight="5dp" android:layout_marginRight="2dp" android:gravity="center"/>
+                <TextView
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="Welcome back! Please log in."
```
Comment (wpride) [inline app/res/layout/screen_login.xml]: No id here - localizable?

File: app/src/org/commcare/dalvik/activities/LoginActivity.java (hunk 6 above)
Comment (dcluna) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: Yes, is there a better way to do it?

File: app/src/org/commcare/dalvik/activities/LoginActivity.java [hunk 10]
```diff
@@ -446,13 +474,25 @@ private void raiseMessage(NotificationMessage message, boolean showTop) {
...
+
         errorBox.setVisibility(View.VISIBLE);
         errorBox.setText(toastText);
         
         Toast.makeText(this,toastText, Toast.LENGTH_LONG).show();
     }
-    
-    
+
+    private void setLoginBoxesColor(int colorId, boolean resolve) {
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: 'resolve' is an uninformative argument name and without method documentation, I have no idea what it is used for?

File: app/src/org/commcare/dalvik/activities/LoginActivity.java (hunk 6 above)
Comment (wpride) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: The other pattern would be having some state Enum that tracks whether we're in error mode or not

File: app/res/layout/screen_login.xml (hunk 9 above)
Comment (ctsims) [inline app/res/layout/screen_login.xml]: @dcluna Agreed. This should have an Id to be localizable

File: app/res/layout/screen_login.xml [hunk 11]
```diff
@@ -51,60 +59,91 @@
...
+                    android:textColor="@color/cc_neutral_color"
+                    android:layout_gravity="center"
+                    android:layout_margin="16dp"
```
Comment (ctsims) [inline app/res/layout/screen_login.xml]: This should be in a common styleset, right? 16dp margins and medium text size for a textview seem like they are likely to be applied broadly?

File: app/res/layout/screen_login.xml (hunk 8 above)
Comment (ctsims) [inline app/res/layout/screen_login.xml]: Probably manual, but should definitely be done

File: app/src/org/commcare/dalvik/activities/LoginActivity.java [hunk 12]
```diff
@@ -97,11 +121,14 @@ public void onCreate(Bundle savedInstanceState) {
...
             }
         }
+
+        final Activity loginAct = this;
         
         login.setOnClickListener(new OnClickListener() {
 
             public void onClick(View arg0) {
                 errorBox.setVisibility(View.GONE);
+                ViewUtil.hideVirtualKeyboard(loginAct);
```
Comment (ctsims) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: You can use the pattern

LoginActivity.this
to get the pointer to the outer class here, which will remove the need for the final declartion above

File: app/res/layout/screen_login.xml (hunk 8 above)
Comment (ctsims) [inline app/res/layout/screen_login.xml]: Note: Luna is doing this in the code below. Can close out this comment

File: app/src/org/commcare/dalvik/activities/LoginActivity.java [hunk 13]
```diff
@@ -74,11 +73,35 @@
...
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+                setLoginBoxesColor(editTextColor, false);
```
Comment (ctsims) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: i'd pull all of these into a helper method in the activity called "setStyleDefault" or something similar to make it more clear what exactly each element is doing and make it triggerable from outside of this context as well.

File: app/src/org/commcare/dalvik/activities/LoginActivity.java [hunk 14]
```diff
@@ -446,13 +474,30 @@ private void raiseMessage(NotificationMessage message, boolean showTop) {
...
     }
-    
-    
+
+    /**
+     * Sets the login boxes (user/pass) to the given color.
+     * @param colorIdOrCode Color ID of resource or color code
+     * @param isColorId Calls getResources.getColor if set, uses the given colorIdOrCode as the color code if not
+     */
+    private void setLoginBoxesColor(int colorIdOrCode, boolean isColorId) {
```
Comment (ctsims) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: This seems a bit over-complicated to me as a signature, honestly. It seems likely that it's worth the hit to the external method to just move the "getResources()" call to the caller and unify what the method does.

File: app/src/org/commcare/dalvik/activities/LoginActivity.java (hunk 6 above)
Comment (dcluna) [inline app/src/org/commcare/dalvik/activities/LoginActivity.java]: Added in latest commit

File: app/res/layout/screen_login.xml (hunk 7 above)
Comment (dcluna) [inline app/res/layout/screen_login.xml]: https://github.com/dimagi/commcare-odk/blob/restyling/login-screen-only-merge/app/src/org/commcare/dalvik/activities/LoginActivity.java#L67

File: app/res/layout/screen_login.xml (hunk 3 above)
Comment (dcluna) [inline app/res/layout/screen_login.xml]: It wraps the CommCare banner, so it's readjusted automatically

**Pull request merged**